cpp/src/round/round.cu

```diff
@@ -0,0 +1,152 @@
+double __device__ generic_modf(double a, double* b) { return modf(a, b); }
+
+template <typename T, typename std::enable_if_t<std::is_signed<T>::value>* = nullptr>
+T __device__ generic_abs(T value)
+{
+  return abs(value);
+}
+
+template <typename T, typename std::enable_if_t<not std::is_signed<T>::value>* = nullptr>
+T __device__ generic_abs(T value)
+{
+  return value;
+}
+
+struct round_fn {
+  template <typename T, typename... Args>
+  std::enable_if_t<not cudf::is_numeric<T>(), std::unique_ptr<column>> operator()(Args&&... args)
+  {
+    CUDF_FAIL("Type not support for cudf::round");
+  }
+
+  template <typename T>
+  std::enable_if_t<cudf::is_floating_point<T>(), std::unique_ptr<column>> operator()(
+    column_view const& input,
+    int32_t decimal_places,
+    cudf::rounding_method method,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream)
+  {
+    auto result   = cudf::make_fixed_width_column(input.type(), input.size());
+    auto out_view = result->mutable_view();
+    T const n     = std::pow(10, std::abs(decimal_places));
+
+    if (decimal_places == 0)
+      thrust::transform(rmm::exec_policy(stream)->on(stream),
+                        input.begin<T>(),
+                        input.end<T>(),
+                        out_view.begin<T>(),
+                        [] __device__(T e) -> T { return generic_round(e); });
+    else if (decimal_places > 0)
+      thrust::transform(rmm::exec_policy(stream)->on(stream),
+                        input.begin<T>(),
+                        input.end<T>(),
+                        out_view.begin<T>(),
+                        [n] __device__(T e) -> T {
+                          T integer_part;
+                          T const fractional_part = generic_modf(e, &integer_part);
+                          return integer_part + generic_round(fractional_part * n) / n;
+                        });
+    else  // decimal_places < 0
+      thrust::transform(rmm::exec_policy(stream)->on(stream),
+                        input.begin<T>(),
+                        input.end<T>(),
+                        out_view.begin<T>(),
+                        [n] __device__(T e) -> T { return generic_round(e / n) * n; });
+
+    return result;
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_integral<T>::value, std::unique_ptr<column>> operator()(
+    column_view const& input,
+    int32_t decimal_places,
+    cudf::rounding_method method,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream)
+  {
+    // only need to handle the case where decimal_places is < zero
+    // integers by definition have no fractional part, so result of "rounding" is a no-op
+    if (decimal_places >= 0) return std::make_unique<cudf::column>(input, stream, mr);
+
+    auto result = cudf::make_fixed_width_column(input.type(), input.size());
+
+    auto out_view = result->mutable_view();
+    auto const n  = static_cast<int64_t>(std::pow(10, -decimal_places));
```

Comment:
Don't use 64-bit values unless absolutely necessary. 32-bit integer arithmetic is much faster on the device, so if T is `int`, it's overkill for `n` to be a 64-bit value, right?

Reply:
Yep, you are totally right. Fixed.